File: src/main/java/siosio/doma/editor/DaoMethodLineMarkerProvider.java
package siosio.doma.editor;

import com.intellij.codeInsight.daemon.RelatedItemLineMarkerInfo;
import com.intellij.codeInsight.daemon.RelatedItemLineMarkerProvider;
import com.intellij.codeInsight.navigation.NavigationGutterIconBuilder;
import com.intellij.openapi.util.IconLoader;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import com.intellij.psi.PsiMethod;
import org.jetbrains.annotations.NotNull;
import siosio.doma.DaoType;
import siosio.doma.DomaBundle;
import siosio.doma.DomaUtils;

import javax.swing.*;
import java.util.Collection;
import java.util.List;

/**
 * DAOのメソッドからSQLファイルへの移動を実現するやつ
 */
public class DaoMethodLineMarkerProvider extends RelatedItemLineMarkerProvider {

    /**
     * SQLファイルのアイコン
     */
    private static final Icon SQL_FILE_ICON = IconLoader.getIcon("/icons/sql.png");

    @Override
    public void collectNavigationMarkers(List<PsiElement> elements,
                                         Collection<? super RelatedItemLineMarkerInfo> result, boolean forNavigation) {
        super.collectNavigationMarkers(elements, result, forNavigation);
    }

    @Override
    protected void collectNavigationMarkers(
            @NotNull PsiElement element,
            Collection<? super RelatedItemLineMarkerInfo> result) {

        if (!(element instanceof PsiMethod)) {
            return;
        }

        VirtualFile file = makeSqlFilePath((PsiMethod) element);
        if (file == null) {
            return;
        }

        PsiFile psiFile = PsiManager.getInstance(element.getProject()).findFile(file);
        NavigationGutterIconBuilder<PsiElement> builder = NavigationGutterIconBuilder
                .create(SQL_FILE_ICON)
                .setTargets(psiFile)
                .setTooltipText(DomaBundle.message("editor.goto-sql-file"));
        result.add(builder.createLineMarkerInfo(element));
    }

    /**
     * メソッドに対応したSQLファイルのパスを取得する。
     *
     * @param method メソッド
     * @return SQLファイルのパス
     */
    private VirtualFile makeSqlFilePath(PsiMethod method) {
        DaoType type = DomaUtils.toDaoType(method);
        if (type == null) {
            return null;
        }
        String slqFilePath = DomaUtils.makeSqlFilePath(method);
        return DomaUtils.findSqlFile(method, slqFilePath);
    }
}

File: src/main/java/siosio/doma/editor/DaoMethodLineMarkerProvider.kt
package siosio.doma.editor

import com.intellij.codeInsight.daemon.*
import com.intellij.codeInsight.navigation.*
import com.intellij.openapi.application.*
import com.intellij.openapi.util.*
import com.intellij.psi.*
import org.jetbrains.kotlin.psi.*
import siosio.doma.*
import siosio.doma.extension.*
import siosio.doma.psi.*

class DaoMethodLineMarkerProvider : RelatedItemLineMarkerProvider() {

    override fun collectNavigationMarkers(element: PsiElement,
                                          result: MutableCollection<in RelatedItemLineMarkerInfo<*>>) {
        
        // テストモードは以下の処理は実行しない
        if (ApplicationManager.getApplication().isUnitTestMode) {
            return
        }

        when (element) {
            is PsiMethod -> PsiDaoMethod(element, DaoType.valueOf(element) ?: return ).findSqlFile()
            is KtNamedFunction -> PsiDaoFunction(element, DaoType.valueOf(element) ?: return).findSqlFile()
            else -> null
        }?.let { 
            element.project.findFile(it)
        }?.let {
            println(SQL_FILE_ICON)
            result.add(
                    NavigationGutterIconBuilder.create(SQL_FILE_ICON)
                            .setTargets(it)
                            .setTooltipText(DomaBundle.message("editor.goto-sql-file"))
                            .createLineMarkerInfo(element.firstChild)
            )
        }
    }
}

private val SQL_FILE_ICON = IconLoader.getIcon("database-solid.svg", DaoMethodLineMarkerProvider::class.java.classLoader)